CMakeLists.txt

```diff
@@ -346,23 +354,47 @@ target_link_libraries(hellouhdm surelog)
-  COMMAND mkdir -p ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/python
-  COMMAND ln -fs ${PROJECT_SOURCE_DIR}/third_party/UVM/ovm-2.1.2 ovm-2.1.2
-  COMMAND ln -fs ${PROJECT_SOURCE_DIR}/third_party/UVM/1800.2-2017-1.0 1800.2-2017-1.0
-  COMMAND ln -fs ${PROJECT_SOURCE_DIR}/third_party/UVM/vmm-1.1.1a vmm-1.1.1a
-  COMMAND rm -rf slpp_all slpp_unit
-  COMMAND cp -f ${PROJECT_SOURCE_DIR}/src/API/builtin.sv ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/sv/builtin.sv
-  COMMAND cp -f ${PROJECT_SOURCE_DIR}/src/API/slSV3_1aPythonListener.py
+  COMMAND echo "       Creating platform independent staging for precompiled packages"
+  COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/sv
+  COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/python
+  COMMAND ${CMAKE_COMMAND} -E remove_directory ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/slpp_all
+  COMMAND ${CMAKE_COMMAND} -E remove_directory ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/slpp_unit
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different ${PROJECT_SOURCE_DIR}/src/API/builtin.sv
+          ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/sv/builtin.sv
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different ${PROJECT_SOURCE_DIR}/src/API/slSV3_1aPythonListener.py
           ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/python/slSV3_1aPythonListener.py
-  COMMAND cp -f ${PROJECT_SOURCE_DIR}/src/API/slformatmsg.py
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different ${PROJECT_SOURCE_DIR}/src/API/slformatmsg.py
           ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/python/slformatmsg.py
-  COMMAND cp -f ${PROJECT_SOURCE_DIR}/src/API/slwaivers.py
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different ${PROJECT_SOURCE_DIR}/src/API/slwaivers.py
           ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/python/slwaivers.py
-  COMMAND echo "       Staging created"
+  COMMAND echo "       Platform independent staging completed"
   WORKING_DIRECTORY "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}")
 
+if (WIN32)
+  add_custom_command(
+    TARGET surelog-bin
+    POST_BUILD
+    COMMAND echo "       Creating platform depdendent staging for precompiled packages"
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_SOURCE_DIR}/third_party/UVM/ovm-2.1.2 ovm-2.1.2
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_SOURCE_DIR}/third_party/UVM/1800.2-2017-1.0 1800.2-2017-1.0
+    COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_SOURCE_DIR}/third_party/UVM/vmm-1.1.1a vmm-1.1.1a
+    COMMAND ${CMAKE_COMMAND} -E copy_if_different
+            ${Python3_RUNTIME_LIBRARY_DIRS}/python3${Python3_VERSION_MINOR}$<$<CONFIG:Debug>:_d>.dll
+            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
+    COMMAND echo "       Platform depdendent staging completed"
+    WORKING_DIRECTORY "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}")
+else()
+  add_custom_command(
+    TARGET surelog-bin
+    POST_BUILD
+    COMMAND echo "       Creating platform depdendent staging for precompiled packages"
+    COMMAND ln -fs ${PROJECT_SOURCE_DIR}/third_party/UVM/ovm-2.1.2 ovm-2.1.2
```

Comment:
Question for @alaindargelas : for simplicity and ease to maintain different platforms, should we maybe just have always the copy_directory version from above ? It means a little wasted disk-space and time for Unix builds, but I think the unifomrity and maintainability is worth it.

Reply:
Ideally, we would do neither copy nor link. We should be able to provide all these paths as inputs and the binary should pick it up. Is there a reason why all of these have to be under one single root folder? Why can't we use the UVM & OVM from in place in third party folder? Sane goes for the scripts as well. Is there a reason why we can't just list them as command line arguments to the tool.

Reply:
Yes good question - @alaindargelas is the right person to answer that.

Reply:
Because we don't want any relative paths of any sort saved in the database. If these files are not under the current dir, you'll have some form of relative path. This is for precompiled packages that should be "portable", meaning, no relative paths. You'd have to perform some hacking to mask the relative path of the files.

Reply:
The copy systematically does not work for me, it would have to copy only if the files are not there already, else the time stamp would change, triggering a recompilation of the pre-compiled  packages (which is what this step is performing). That is an expensive task that should only be performed when the grammar files change in one's build enviroment.

Reply:
Yes that is going to be a problem at the moment since copy_directory command always overwrite the destination contents regardless. Here's a custom implementation - an attempt to implement update_directory command - which can be customized to our needs.

https://bravenewmethod.com/2017/06/18/update_directory-command-for-cmake/

Possibly an exercise for some other day!